File: server/internal/wealth/db/real_estate.go
package wealthdb

import (
	"context"
	"database/sql"
	"fmt"

	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
	"tallyo/internal/wealth"
)

// CreateRealEstate links a home and its initial valuation in one transaction.
func (s *Store) CreateRealEstate(ctx context.Context, in wealth.CreateRealEstateInput) (*wealth.CreateRealEstateResult, error) {
	owner, err := s.OwnerByID(ctx, in.OwnerID)
	if err != nil {
		return nil, fmt.Errorf("lookup owner: %w", err)
	}
	if owner == nil {
		return nil, fmt.Errorf("unknown owner id %d", in.OwnerID)
	}

	name := in.Label
	if name == "" {
		name = defaultRealEstateName(in.Details)
	}
	externalID, err := dbutil.RandomID()
	if err != nil {
		return nil, err
	}

	additional, err := marshalAdditional(realEstateAdditional(in.Details))
	if err != nil {
		return nil, err
	}
	var result *wealth.CreateRealEstateResult
	if err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		assetID, err := q.InsertRealEstateAsset(ctx, dbgen.InsertRealEstateAssetParams{
			Identifier: externalID,
			Name:       &name,
			Additional: additional,
		})
		if err != nil {
			return fmt.Errorf("insert real estate asset: %w", err)
		}
		connID, err := q.UpsertConnection(ctx, dbgen.UpsertConnectionParams{
			SourceTable: string(wealth.SourceTableAssets),
			SourceID:    assetID,
			Name:        &name,
			OwnerID:     in.OwnerID,
		})
		if err != nil {
			return fmt.Errorf("insert connection: %w", err)
		}
		accountID, err := q.InsertLinkedAccount(ctx, dbgen.InsertLinkedAccountParams{
			ExternalID:   externalID,
			ConnectionID: &connID,
			OwnerID:      in.OwnerID,
			Name:         name,
			Type:         string(model.AccountTypeProperty),
		})
		if err != nil {
			return fmt.Errorf("insert real estate account: %w", err)
		}
		home := wealth.RealEstate{
			AssetID:      assetID,
			AccountID:    accountID,
			ConnectionID: connID,
			OwnerID:      in.OwnerID,
			Name:         name,
			Details:      in.Details,
		}
		if in.InitialValuation != nil {
			if err := s.persistSnapshot(ctx, q, in.InitialValuation.Persist(home)); err != nil {
				return fmt.Errorf("persist real estate valuation: %w", err)
			}
		}
		conn := &model.Connection{
			ID:          model.New(model.GlobalIDConnection, connID),
			Name:        &name,
			Owner:       owner,
			IsActive:    true,
			SourceTable: string(wealth.SourceTableAssets),
			SourceID:    assetID,
		}
		account := &model.Account{
			ID:         model.New(model.GlobalIDAccount, accountID),
			Connection: conn,
			Owner:      owner,
			Name:       name,
			Type:       model.AccountTypeProperty,
			Manual:     false,
		}
		result = &wealth.CreateRealEstateResult{Connection: conn, Account: account}
		return nil
	}); err != nil {
		return nil, err
	}

	return result, nil
}

// UpdateRealEstate updates details and an optional valuation in one transaction.
func (s *Store) UpdateRealEstate(ctx context.Context, in wealth.UpdateRealEstateInput) (int64, error) {
	var accountID int64
	if err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		re, err := realEstateByConnectionID(ctx, q, in.ConnectionID)
		if err != nil {
			return fmt.Errorf("lookup real estate: %w", err)
		}
		if re == nil {
			return fmt.Errorf("%w for connection %d", wealth.ErrRealEstateNotFound, in.ConnectionID)
		}
		if err := updateRealEstateDetails(ctx, q, *re, in.Name, in.Details); err != nil {
			return err
		}
		if in.Valuation != nil {
			if err := s.persistSnapshot(ctx, q, in.Valuation.Persist(*re)); err != nil {
				return fmt.Errorf("persist real estate valuation: %w", err)
			}
		}
		accountID = re.AccountID
		return nil
	}); err != nil {
		return 0, err
	}
	return accountID, nil
}

func updateRealEstateDetails(
	ctx context.Context,
	q *dbgen.Queries,
	re wealth.RealEstate,
	name *string,
	details wealth.RealEstateDetails,
) error {
	incoming, err := marshalAdditional(realEstateAdditional(details))
	if err != nil {
		return err
	}
	if name != nil {
		if err := q.UpdateAssetBase(ctx, dbgen.UpdateAssetBaseParams{Name: name, ID: re.AssetID}); err != nil {
			return fmt.Errorf("update asset name: %w", err)
		}
		if _, err := q.UpdateAccountFields(ctx, dbgen.UpdateAccountFieldsParams{SetName: true, Name: *name, ID: re.AccountID}); err != nil {
			return fmt.Errorf("update account name: %w", err)
		}
		if _, err := q.UpsertConnection(ctx, dbgen.UpsertConnectionParams{
			SourceTable: string(wealth.SourceTableAssets),
			SourceID:    re.AssetID,
			Name:        name,
			OwnerID:     re.OwnerID,
		}); err != nil {
			return fmt.Errorf("update connection name: %w", err)
		}
	}
	if incoming == nil {
		return nil
	}
	existing, err := assetAdditionalByID(ctx, q, re.AssetID)
	if err != nil {
		return fmt.Errorf("read real estate additional: %w", err)
	}
	merged, err := mergeAdditional(model.AssetTypeRealEstate, existing, incoming)
	if err != nil {
		return err
	}
	if err := q.UpdateAssetAdditional(ctx, dbgen.UpdateAssetAdditionalParams{Additional: merged, ID: re.AssetID}); err != nil {
		return fmt.Errorf("update real estate additional: %w", err)
	}
	return nil
}

// RealEstateByConnectionID returns the linked home for a given connection ID.
func (s *Store) RealEstateByConnectionID(ctx context.Context, connectionID int64) (*wealth.RealEstate, error) {
	return realEstateByConnectionID(ctx, s.q, connectionID)
}

func realEstateByConnectionID(ctx context.Context, q *dbgen.Queries, connectionID int64) (*wealth.RealEstate, error) {
	rows, err := q.ActiveRealEstate(ctx, dbgen.ActiveRealEstateParams{ConnectionID: &connectionID})
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	re, err := realEstateFromSQLRow(rows[0])
	if err != nil {
		return nil, err
	}
	return &re, nil
}

// ActiveRealEstate returns all linked homes with an open account, used for
// periodic valuation sync. Homes on closed accounts are excluded because the
// shared snapshot sink rejects writes for them.
func (s *Store) ActiveRealEstate(ctx context.Context) ([]wealth.RealEstate, error) {
	rows, err := s.q.ActiveRealEstate(ctx, dbgen.ActiveRealEstateParams{OpenOnly: true})
	if err != nil {
		return nil, fmt.Errorf("list real estate: %w", err)
	}
	return u.MapErr(rows, realEstateFromSQLRow)
}

// DeleteRealEstate removes the account, snapshots, connection, and the
// REAL_ESTATE asset for a connection, in one transaction.
func (s *Store) DeleteRealEstate(ctx context.Context, connectionID int64) error {
	return s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		re, err := realEstateByConnectionID(ctx, q, connectionID)
		if err != nil {
			return err
		}
		if re == nil {
			return fmt.Errorf("real estate for connection %d not found", connectionID)
		}
		if err := q.DeleteAccountsByConnection(ctx, dbgen.DeleteAccountsByConnectionParams{ConnectionID: &connectionID}); err != nil {
			return fmt.Errorf("delete real estate accounts: %w", err)
		}
		if err := q.DeleteConnectionByID(ctx, dbgen.DeleteConnectionByIDParams{ID: connectionID}); err != nil {
			return fmt.Errorf("delete real estate connection: %w", err)
		}
		if err := q.DeleteAssetByID(ctx, dbgen.DeleteAssetByIDParams{ID: re.AssetID}); err != nil {
			return fmt.Errorf("delete real estate asset: %w", err)
		}
		return nil
	})
}

func realEstateFromSQLRow(row dbgen.ActiveRealEstateRow) (wealth.RealEstate, error) {
	var details RealEstateAdditional
	if err := unmarshalAdditional(row.Additional, &details); err != nil {
		return wealth.RealEstate{}, err
	}
	return wealth.RealEstate{
		AssetID:      row.AssetID,
		AccountID:    row.AccountID,
		ConnectionID: row.ConnectionID,
		OwnerID:      row.OwnerID,
		Name:         row.Name,
		Details: wealth.RealEstateDetails{
			Street:   details.Street,
			City:     details.City,
			State:    details.State,
			Zip:      details.Zip,
			HomeType: details.HomeType,
		},
	}, nil
}

func defaultRealEstateName(re wealth.RealEstateDetails) string {
	if re.Street != nil && *re.Street != "" {
		return *re.Street
	}
	return "Home"
}

File: server-rs/src/wealth/store/real_estate.rs
use anyhow::{Context, Result, anyhow};
use rand::RngCore;
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    accounts::{
        Account, AccountType, ConnectionRecord, SourceTable,
        store::{account_by_id, connection_by_id, owner_by_id},
    },
    database::{self, queries},
    wealth::{CreateRealEstate, RealEstate, RealEstateDetails, UpdateRealEstate},
};

use super::{
    asset_additional::{RealEstateAdditional, merge_additional, real_estate_additional},
    assets::asset_additional_by_id,
    snapshots::persist_snapshot_in_transaction,
};

pub async fn create_real_estate(pool: &SqlitePool, input: CreateRealEstate) -> Result<(ConnectionRecord, Account)> {
    if owner_by_id(pool, input.owner_id)
        .await
        .context("lookup owner")?
        .is_none()
    {
        return Err(anyhow!("unknown owner id {}", input.owner_id));
    }
    let name = default_name(&input.label, &input.details);
    let external_id = random_id();
    let additional = details_json(&input.details)?;
    let (connection_id, account_id) = database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let asset_id = queries::insert_real_estate_asset(
                &mut **transaction,
                queries::InsertRealEstateAssetParams {
                    identifier: &external_id,
                    name: Some(&name),
                    additional: additional.as_deref(),
                },
            )
            .await
            .context("insert real estate asset")?
            .id;
            let connection_id = queries::upsert_connection(
                &mut **transaction,
                queries::UpsertConnectionParams {
                    source_table: &SourceTable::Assets.to_string(),
                    source_id: asset_id,
                    name: Some(&name),
                    owner_id: input.owner_id,
                },
            )
            .await
            .context("insert real estate connection")?
            .id;
            let account_id = queries::insert_linked_account(
                &mut **transaction,
                queries::InsertLinkedAccountParams {
                    external_id: &external_id,
                    connection_id: Some(connection_id),
                    owner_id: input.owner_id,
                    name: &name,
                    r#type: &AccountType::Property.to_string(),
                },
            )
            .await
            .context("insert real estate account")?
            .id;
            if let Some(valuation) = input.initial_valuation {
                persist_snapshot_in_transaction(
                    transaction,
                    valuation.persist(&RealEstate {
                        asset_id,
                        account_id,
                        connection_id,
                        owner_id: input.owner_id,
                        name: name.clone(),
                        details: input.details,
                    }),
                )
                .await
                .context("persist real estate valuation")?;
            }
            Ok((connection_id, account_id))
        })
    })
    .await?;
    let connection = connection_by_id(pool, connection_id)
        .await?
        .ok_or_else(|| anyhow!("connection {connection_id} not found"))?;
    let account = account_by_id(pool, account_id)
        .await?
        .ok_or_else(|| anyhow!("account {account_id} not found"))?;
    Ok((connection, account))
}

pub async fn update_real_estate(pool: &SqlitePool, input: UpdateRealEstate) -> Result<i64> {
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let real_estate = real_estate_by_connection_id(&mut **transaction, input.connection_id)
                .await?
                .ok_or_else(|| anyhow!("real estate for connection {} not found", input.connection_id))?;
            update_details(transaction, &real_estate, input.name.as_deref(), &input.details).await?;
            if let Some(valuation) = input.valuation {
                persist_snapshot_in_transaction(transaction, valuation.persist(&real_estate))
                    .await
                    .context("persist real estate valuation")?;
            }
            Ok(real_estate.account_id)
        })
    })
    .await
}

pub async fn real_estate_by_connection_id(
    executor: impl Executor<'_, Database = Sqlite>,
    connection_id: i64,
) -> Result<Option<RealEstate>> {
    queries::active_real_estate(
        executor,
        queries::ActiveRealEstateParams {
            connection_id: Some(connection_id),
            ..Default::default()
        },
    )
    .await?
    .into_iter()
    .next()
    .map(real_estate_from_row)
    .transpose()
}

pub async fn active_real_estate(executor: impl Executor<'_, Database = Sqlite>) -> Result<Vec<RealEstate>> {
    queries::active_real_estate(
        executor,
        queries::ActiveRealEstateParams {
            open_only: true,
            ..Default::default()
        },
    )
    .await?
    .into_iter()
    .map(real_estate_from_row)
    .collect()
}

pub async fn delete_real_estate(pool: &SqlitePool, connection_id: i64) -> Result<()> {
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let real_estate = real_estate_by_connection_id(&mut **transaction, connection_id)
                .await?
                .ok_or_else(|| anyhow!("real estate for connection {connection_id} not found"))?;
            queries::delete_accounts_by_connection(
                &mut **transaction,
                queries::DeleteAccountsByConnectionParams {
                    connection_id: Some(connection_id),
                },
            )
            .await
            .context("delete real estate accounts")?;
            queries::delete_connection_by_id(
                &mut **transaction,
                queries::DeleteConnectionByIdParams { id: connection_id },
            )
            .await
            .context("delete real estate connection")?;
            queries::delete_asset_by_id(
                &mut **transaction,
                queries::DeleteAssetByIdParams {
                    id: real_estate.asset_id,
                },
            )
            .await
            .context("delete real estate asset")
        })
    })
    .await
}

async fn update_details(
    transaction: &mut sqlx::Transaction<'_, Sqlite>,
    real_estate: &RealEstate,
    name: Option<&str>,
    details: &RealEstateDetails,
) -> Result<()> {
    if let Some(name) = name {
        queries::update_asset_base(
            &mut **transaction,
            queries::UpdateAssetBaseParams {
                name: Some(name),
                id: real_estate.asset_id,
                ..Default::default()
            },
        )
        .await
        .context("update real estate asset name")?;
        queries::update_account_fields(
            &mut **transaction,
            queries::UpdateAccountFieldsParams {
                set_name: true,
                name,
                id: real_estate.account_id,
                ..Default::default()
            },
        )
        .await
        .context("update real estate account name")?;
        queries::upsert_connection(
            &mut **transaction,
            queries::UpsertConnectionParams {
                source_table: &SourceTable::Assets.to_string(),
                source_id: real_estate.asset_id,
                name: Some(name),
                owner_id: real_estate.owner_id,
            },
        )
        .await
        .context("update real estate connection name")?;
    }
    let Some(incoming) = details_json(details)? else {
        return Ok(());
    };
    let current = asset_additional_by_id(&mut **transaction, real_estate.asset_id).await?;
    let merged = merge_additional(
        crate::schema::AssetType::RealEstate,
        current.as_deref(),
        Some(&incoming),
    )?;
    queries::update_asset_additional(
        &mut **transaction,
        queries::UpdateAssetAdditionalParams {
            additional: merged.as_deref(),
            id: real_estate.asset_id,
        },
    )
    .await
    .context("update real estate additional")
}

fn real_estate_from_row(row: queries::ActiveRealEstateRow) -> Result<RealEstate> {
    let details = row
        .additional
        .filter(|additional| !additional.is_empty())
        .map(|additional| serde_json::from_str::<RealEstateAdditional>(&additional))
        .transpose()
        .context("decode real estate additional")?
        .unwrap_or_default();
    Ok(RealEstate {
        asset_id: row.asset_id,
        account_id: row.account_id,
        connection_id: row.connection_id,
        owner_id: row.owner_id,
        name: row.name,
        details: RealEstateDetails {
            street: details.street,
            city: details.city,
            state: details.state,
            zip: details.zip,
            home_type: details.home_type,
        },
    })
}

fn default_name(label: &str, details: &RealEstateDetails) -> String {
    if label.is_empty() {
        details
            .street
            .as_deref()
            .filter(|street| !street.is_empty())
            .unwrap_or("Home")
            .to_owned()
    } else {
        label.to_owned()
    }
}

fn details_json(details: &RealEstateDetails) -> Result<Option<String>> {
    let json = serde_json::to_string(&real_estate_additional(details))?;
    Ok((json != "{}").then_some(json))
}

fn random_id() -> String {
    let mut bytes = [0_u8; 16];
    rand::rng().fill_bytes(&mut bytes);
    hex::encode(bytes)
}
